From 4a5efa4af9074eda5146ba72f7015dabdfc0b37f Mon Sep 17 00:00:00 2001
From: kongweibin <kongweibin2@huawei.com>
Date: Sat, 14 Oct 2023 14:42:36 +0800
Subject: [PATCH 8/8] bpf: Two helper functions are introduced to parse
 user-mode messages.

User messages need to be parsed when the bpf link establishment
is delayed. In this case, two helpers need to be introduced. One is
to parse user-mode messages, and the other is to obtain the parsed
data.

Signed-off-by: kongweibin <kongweibin2@huawei.com>
---
 include/uapi/linux/bpf.h       | 21 ++++++++++++++++++
 net/core/filter.c              | 40 ++++++++++++++++++++++++++++++++++
 scripts/bpf_helpers_doc.py     |  1 +
 tools/include/uapi/linux/bpf.h | 21 ++++++++++++++++++
 4 files changed, 83 insertions(+)

diff --git a/include/uapi/linux/bpf.h b/include/uapi/linux/bpf.h
index 703c6f124..8883598eb 100644
--- a/include/uapi/linux/bpf.h
+++ b/include/uapi/linux/bpf.h
@@ -3908,6 +3908,20 @@ union bpf_attr {
  *		An integer less than, equal to, or greater than zero
  *		if the first **s1_sz** bytes of **s1** is found to be
  *		less than, to match, or be greater than **s2**.
+ *
+ * long bpf_parse_header_msg(struct bpf_mem_ptr *msg)
+ * 	Description
+ * 		Parses the content of the msg. User can use `parse_protocol_func`
+ * 		to define the parse function.
+ * 	Return
+ * 		User-defined return value.
+ *
+ * void *bpf_get_msg_header_element(void *name)
+ * 	Description
+ * 		Reads the content of the parsed msg. User can use
+ * 		`get_protocol_element_func` to define the content.
+ * 	Return
+ * 		Contains a pointer to the data and the length of the data.
  */
 #define __BPF_FUNC_MAPPER(FN)		\
 	FN(unspec),			\
@@ -4085,6 +4099,8 @@ union bpf_attr {
 	FN(strnstr),			\
 	FN(strnlen),			\
 	FN(_strncmp),			\
+	FN(parse_header_msg),		\
+	FN(get_msg_header_element),	\
 	/* */
 
 /* integer value in 'imm' field of BPF_CALL instruction selects which helper
@@ -5222,6 +5238,11 @@ struct btf_ptr {
 	__u32 flags;		/* BTF ptr flags; unused at present. */
 };
 
+struct bpf_mem_ptr {
+	void *ptr;
+	__u32 size;
+};
+
 /*
  * Flags to control bpf_snprintf_btf() behaviour.
  *     - BTF_F_COMPACT: no formatting around type information
diff --git a/net/core/filter.c b/net/core/filter.c
index 6a0fdc5ce..ee9aa48d3 100644
--- a/net/core/filter.c
+++ b/net/core/filter.c
@@ -6991,6 +6991,42 @@ static const struct bpf_func_proto bpf_sock_ops_reserve_hdr_opt_proto = {
 	.arg3_type	= ARG_ANYTHING,
 };
 
+typedef int (*bpf_parse_protocol_func)(struct bpf_mem_ptr* msg);
+bpf_parse_protocol_func parse_protocol_func = NULL;
+EXPORT_SYMBOL(parse_protocol_func);
+
+typedef void* (*bpf_get_protocol_element_func)(char *key);
+bpf_get_protocol_element_func get_protocol_element_func = NULL;
+EXPORT_SYMBOL(get_protocol_element_func);
+
+BPF_CALL_1(bpf_parse_header_msg, struct bpf_mem_ptr *, msg)
+{
+	if (!parse_protocol_func)
+		return -ENOTSUPP;
+	return parse_protocol_func(msg);
+}
+
+static const struct bpf_func_proto bpf_parse_header_msg_proto = {
+	.func		= bpf_parse_header_msg,
+	.gpl_only	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_ANYTHING,
+};
+
+BPF_CALL_1(bpf_get_msg_header_element, char *, key)
+{
+	if (!get_protocol_element_func)
+		return -ENOTSUPP;
+	return get_protocol_element_func(key);
+}
+
+static const struct bpf_func_proto bpf_get_msg_header_element_proto = {
+	.func		= bpf_get_msg_header_element,
+	.gpl_only	= false,
+	.ret_type	= RET_PTR_TO_ALLOC_MEM_OR_NULL,
+	.arg1_type	= ARG_ANYTHING,
+};
+
 #endif /* CONFIG_INET */
 
 bool bpf_helper_changes_pkt_data(void *func)
@@ -7419,6 +7455,10 @@ sock_ops_func_proto(enum bpf_func_id func_id, const struct bpf_prog *prog)
 		return &bpf_sock_ops_reserve_hdr_opt_proto;
 	case BPF_FUNC_tcp_sock:
 		return &bpf_tcp_sock_proto;
+	case BPF_FUNC_parse_header_msg:
+		return &bpf_parse_header_msg_proto;
+	case BPF_FUNC_get_msg_header_element:
+		return &bpf_get_msg_header_element_proto;
 #endif /* CONFIG_INET */
 	case BPF_FUNC_is_local_ipaddr:
 		return &bpf_is_local_ipaddr_proto;
diff --git a/scripts/bpf_helpers_doc.py b/scripts/bpf_helpers_doc.py
index fc51d6f0d..74cc376b0 100755
--- a/scripts/bpf_helpers_doc.py
+++ b/scripts/bpf_helpers_doc.py
@@ -496,6 +496,7 @@ class PrinterHelpers(Printer):
             'struct cpumask_op_args',
             'struct sched_migrate_ctx',
             'struct sched_affine_ctx',
+            'struct bpf_mem_ptr',
     }
     mapped_types = {
             'u8': '__u8',
diff --git a/tools/include/uapi/linux/bpf.h b/tools/include/uapi/linux/bpf.h
index 703c6f124..8883598eb 100644
--- a/tools/include/uapi/linux/bpf.h
+++ b/tools/include/uapi/linux/bpf.h
@@ -3908,6 +3908,20 @@ union bpf_attr {
  *		An integer less than, equal to, or greater than zero
  *		if the first **s1_sz** bytes of **s1** is found to be
  *		less than, to match, or be greater than **s2**.
+ *
+ * long bpf_parse_header_msg(struct bpf_mem_ptr *msg)
+ * 	Description
+ * 		Parses the content of the msg. User can use `parse_protocol_func`
+ * 		to define the parse function.
+ * 	Return
+ * 		User-defined return value.
+ *
+ * void *bpf_get_msg_header_element(void *name)
+ * 	Description
+ * 		Reads the content of the parsed msg. User can use
+ * 		`get_protocol_element_func` to define the content.
+ * 	Return
+ * 		Contains a pointer to the data and the length of the data.
  */
 #define __BPF_FUNC_MAPPER(FN)		\
 	FN(unspec),			\
@@ -4085,6 +4099,8 @@ union bpf_attr {
 	FN(strnstr),			\
 	FN(strnlen),			\
 	FN(_strncmp),			\
+	FN(parse_header_msg),		\
+	FN(get_msg_header_element),	\
 	/* */
 
 /* integer value in 'imm' field of BPF_CALL instruction selects which helper
@@ -5222,6 +5238,11 @@ struct btf_ptr {
 	__u32 flags;		/* BTF ptr flags; unused at present. */
 };
 
+struct bpf_mem_ptr {
+	void *ptr;
+	__u32 size;
+};
+
 /*
  * Flags to control bpf_snprintf_btf() behaviour.
  *     - BTF_F_COMPACT: no formatting around type information
-- 
2.33.0

